### ISSUE TITLE
Buffered messages are correlated out of order

### ISSUE THREAD

**Issue by saig0** [zeebe-io/zeebe#3397]
**Describe the bug**
If a workflow instance enters multiple message catch events (e.g. event-based gateway with two message catch events, or two message boundary events on an activity) and messages are buffered for these catch events then it can happen that a message is correlated which was published after another one. 

So, the messages are not correlated in the order as they are published. Even, if they are published on the same partition (i.e. same correlation key).

**To Reproduce**

![image](https://user-images.githubusercontent.com/4305769/69228433-d9262b00-0b83-11ea-8860-097d238331b0.png)

* deploy a workflow with an event-based gateway an two message catch events
* publish a message with name "A" (1)
* publish a message with name "B" (2)
* publish a message with name "A" (3)
* publish a message with name "B" (4)
* create one workflow instance for each message (and wait until the gateway is left)
* verify that the messages are correlated in the right order (1, 2, 3, 4)

**Expected behavior**
The messages are correlated in the same order as they are published. If messages are published to different partitions, the order may be different. 

**Log/Stacktrace**

<details><summary>Test Case</summary>
 <p>

```
@Test
  public void shouldCorrelateMessagesInOrder() {
    // given
    engine
        .deployment()
        .withXmlResource(
            Bpmn.createExecutableProcess("wf")
                .startEvent()
                .eventBasedGateway("gw")
                .intermediateCatchEvent(
                    "catch-a", c -> c.message(m -> m.name("a").zeebeCorrelationKey("key")))
                .endEvent()
                .moveToLastGateway()
                .intermediateCatchEvent(
                    "catch-b", c -> c.message(m -> m.name("b").zeebeCorrelationKey("key")))
                .endEvent()
                .done())
        .deploy();

    engine
        .message()
        .withName("a")
        .withCorrelationKey("key-1")
        .withVariables(Map.of("x", 1))
        .publish();

    engine
        .message()
        .withName("b")
        .withCorrelationKey("key-1")
        .withVariables(Map.of("x", 2))
        .publish();

    engine
        .message()
        .withName("a")
        .withCorrelationKey("key-1")
        .withVariables(Map.of("x", 3))
        .publish();

    engine
        .message()
        .withName("b")
        .withCorrelationKey("key-1")
        .withVariables(Map.of("x", 4))
        .publish();

    // when
    IntStream.range(0, 4)
        .forEach(
            i -> {
              final var workflowInstanceKey =
                  engine
                      .workflowInstance()
                      .ofBpmnProcessId("wf")
                      .withVariable("key", "key-1")
                      .create();

              RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_COMPLETED)
                  .withWorkflowInstanceKey(workflowInstanceKey)
                  .filterRootScope()
                  .await();
            });

    // then
    assertThat(RecordingExporter.variableRecords().withName("x").limit(4))
        .extracting(r -> r.getValue().getValue())
        .containsExactly("1", "2", "3", "4");
  }
```

</p>
</details>

Same problem as #2459 but for a different situation. 

**Environment:**
 - OS: [e.g. Linux]
 - Zeebe Version: 0.21.0
 - Configuration: [e.g. exporters etc.]

**Comment by Zelldon** [comment]
@saig0 is this a duplicate of #2459 ?

**Comment by saig0** [comment]
@Zelldon nope. The issues describe different scenarios. But both are related to messages that may be correlated out of order.

**Comment by marcosgvieira** [comment]
### 🧭 Triage Assessment

| Field | Value |
|------|-------|
| **Work type** | BUG |
| **Severity** | 🔥 High |
| **Likelihood** | ⚠️ Mid |
| **Confidence** | 0.78 |

---

### 📝 Rationale

Issue describes incorrect Zeebe engine behavior where buffered messages for multiple message catch events on the same correlation key can be correlated out of publication order, violating expected BPMN semantics and potentially leading to wrong business outcomes. This is engine execution logic, which falls under Core Features (Zeebe Engine and execution) per Team Boundaries, and the issue is still open and explicitly stated not to be a duplicate of #2459. Impact on production usage can be significant with no obvious simple workaround (HIGH severity), while the occurrence is tied to specific modeling patterns and buffering conditions (MID likelihood).

---

🤖 _This issue was automatically triaged by an AI-based system as part of the Core Features triage workflow._

✋ **Want a human review?**
If you believe this assessment is incorrect or needs adjustment, feel free to update the details or request a manual triage by adding the **`triage:manual`** label.

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Backlog  
**To:** Ready  

**Reason:** The issue already has an assignee.

**Comment by remcowesterhoud** [comment]
I looked a bit into this. As expected (and raised before) this is far from trivial to solve. Previously estimated effort was [3-6 months](https://jira.camunda.com/browse/SUPPORT-22315?focusedCommentId=372212&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372212).

First things first, the current behavior is expected. Therefore I don't consider this to be a bug. The linked support case is also a feature request. I will change this issue accordingly.

### Why is this so complicated?

FIFO for messages is difficult for multiple reasons:
1. We are working with a distributed system with multiple partitions running independent from eachother. Message A could arrive on partition 1 where message B arrives on partition 2. Partition 1 knows nothing about the message on partition 2 and vice versa. As a result we cannot apply an ordering here. Both partitions send them to the partition where PI lives (partition 3). Partition health, network latency/issues, etc. all play part in when a message actually gets forwarded to partition 3.
2. This issue also describes FIFO across Process Instances. Each PI could live on a different partition and doesn't know anything about the other PIs.

### What can we do?

At the moment not much. There is a workaround described in the linked support case where the published messages are stored as variables within the process instance.

### Why is it not a bug?

Our documentation mentions FIFO only for [messages correlated to the same subscription](https://docs.camunda.io/docs/components/concepts/messages/#message-buffering).

> When a subscription is opened, it polls the buffer for a proper message. If a proper message exists, it is correlated to the corresponding process instance. In case multiple messages match to the subscription, the first published message is correlated (like a FIFO queue).

This is true. The partition the message arrives at is determined by the correlation key. Messages with the same name and same correlation key are handled FIFO.

### Random thought on how we could fix this

We need to properly revisit the messaging flow if we want to support this. 

⚠️ **I didn't think this option through exhaustively!**

The first option that comes to my mind is sending all message through the same partition. This partition could take care of ensuring they get send to the proper partition in FIFO rodering. This has significant downsides though:

1. Unfair distriution of message publishing. They'd all arrive on the same partition. This also means this one partition becomes a bottleneck for issues if it's broken. We already have this pattern for other commands (deployment, identity CRUD) where all the commands go through partition 1. Ideally we don't add more.
2. This significantly reduces the throughput of messaging. Partiton 1 would send a message to partition X. It would have to wait for acknowledgement of partition X before it can send the next message. This also brings the risk of blocking messages entirely when there's a bug, increasing the blast radius from a single PI now to all PIs that use messages.
